larpandoracontent/LArVertex/MvaVertexSelectionAlgorithm.cc

```diff
@@ -680,19 +687,30 @@ void MvaVertexSelectionAlgorithm<T>::GetBestVertex(const VertexVector &vertexVec
     const MCParticleList *pMCParticleList(nullptr);
     PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, m_mcParticleListName, pMCParticleList));
 
-    // Obtain vector: true neutrinos
-    MCParticleVector mcNeutrinoVector;
-    LArMCParticleHelper::GetTrueNeutrinos(pMCParticleList, mcNeutrinoVector);
+    // Obtain vector: true targets
+    MCParticleVector mcTargetVector;
+    LArMCParticleHelper::GetTrueNeutrinos(pMCParticleList, mcTargetVector);
+
+    if (m_testBeamMode)
+    {
+        LArMCParticleHelper::GetTrueTestBeamParticles(pMCParticleList, mcTargetVector);
+    }
+    else
+    {
+        LArMCParticleHelper::GetTrueNeutrinos(pMCParticleList, mcTargetVector);
+    }
 
     for (const Vertex *const pVertex : vertexVector)
     {
         float mcVertexDr(std::numeric_limits<float>::max());
-        for (const MCParticle *const pMCNeutrino : mcNeutrinoVector)
+        for (const MCParticle *const pMCTarget : mcTargetVector)
         {
-            const CartesianVector mcNeutrinoPosition(pMCNeutrino->GetEndpoint().GetX() + m_mcVertexXCorrection, pMCNeutrino->GetEndpoint().GetY(),
-                pMCNeutrino->GetEndpoint().GetZ());
+            const CartesianVector mcTargetPosition((m_testBeamMode && std::fabs(pMCTarget->GetParticleId()) == 11) ? pMCTarget->GetVertex() :
```

Comment:
This could be a const reference (relatively small gain here, but probably worthwhile)